File: Services/Concrete/Excel/Sheets/Sheet.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Services.Concrete.Excel.Sheets
{
    internal abstract class Sheet
    {
        private readonly Workbook _workbook;
        protected abstract string GetName();
        protected abstract string[] GetColumnNames();
        public abstract void Generate();

        protected Sheet(Workbook workbook)
        {
            _workbook = workbook;
            _workbook.AddSheet(GetName());
            if (GetColumnNames().Length > 0)
            {
                _workbook.AddRowToSheet(GetName(), GetColumnNames().ToList<object>());
            }
        }

        protected void WriteRow(List<object> cells)
        {
            if (GetColumnNames().Length > 0 && cells.Count != GetColumnNames().Length)
            {
                throw new Exception($@"Expected {GetColumnNames().Length} cells, got {cells.Count}");
            }

            _workbook.AddRowToSheet(GetName(), cells);
        }
    }
}

File: src/node/xlsx/sheet.ts
import type { CellValue } from './cell-value';
import type { Column } from './column';
import type { Workbook } from './workbook';

export abstract class Sheet<RowType> {
  private workbook: Workbook;
  protected abstract getName(): string;
  protected abstract getColumns(): Column<RowType>[];
  protected abstract generate(): void | Promise<void>;

  public constructor(workbook: Workbook) {
    this.workbook = workbook;
    this.workbook.addSheet(this.getName());
    const columnNames = this.getColumns().map((column) => column.name);
    if (columnNames.length > 0) {
      this.workbook.addRowToSheet(this.getName(), columnNames);
    }
  }

  protected writeRow(row: RowType) {
    const cells: CellValue[] = [];
    for (const column of this.getColumns()) {
      cells.push(column.cellFormatter(row));
    }

    this.workbook.addRowToSheet(this.getName(), cells);
  }

  protected writeCells(cells: CellValue[]) {
    this.workbook.addRowToSheet(this.getName(), cells);
  }
}
